tests/foreman/api/test_repository.py

```diff
@@ -1453,6 +1453,112 @@ def test_negative_synchronize_private_registry_no_passwd(self):
         self.assertIn("422", str(excinfo.exception))
         self.assertIn("Unprocessable Entity", str(excinfo.exception))
 
+    @run_only_on('sat')
+    @tier2
+    @upgrade
+    def test_positive_synchronize_docker_repo_with_tags_whitelist(self):
+        """Check if only whitelisted tags are synchronized
+
+        :id: abd584ef-f616-49d8-ab30-ae32e4e8a685
+
+        :expectedresults: Only whitelisted tag is synchronized
+        """
+        tags = ['latest']
+        product = entities.Product(organization=self.org).create()
+        repo = entities.Repository(
+                content_type='docker',
+                docker_tags_whitelist=tags,
+                docker_upstream_name='alpine',
+                name=gen_string('alphanumeric', 10),
+                product=product,
+                url=DOCKER_REGISTRY_HUB,
+        ).create()
+        repo.sync()
+        repo = repo.read()
+        [self.assertIn(tag, repo.docker_tags_whitelist) for tag in tags]
+        self.assertEqual(repo.content_counts['docker_tag'], 1)
+
+    @run_only_on('sat')
+    @tier2
+    def test_positive_synchronize_docker_repo_set_tags_later(self):
+        """Verify that adding tags whitelist after synchronizing repository
+        doesn't affect already synchronized content
+
+        :id: 6838e152-5fd9-4f25-ae04-67760571f6ba
+
+        :expectedresults: Non-whitelisted tags are not removed
+        """
+        tags = ['latest']
+        product = entities.Product(organization=self.org).create()
+        repo = entities.Repository(
+                content_type='docker',
+                docker_upstream_name='alpine',
+                name=gen_string('alphanumeric', 10),
+                product=product,
+                url=DOCKER_REGISTRY_HUB,
+        ).create()
+        repo.sync()
+        repo = repo.read()
+        self.assertIsNone(repo.docker_tags_whitelist)
+        self.assertGreaterEqual(repo.content_counts['docker_tag'], 2)
+
+        repo.docker_tags_whitelist = tags
+        repo.update(['docker_tags_whitelist'])
+        repo.sync()
```

Comment:
Your docstring saying:
```
Verify that adding tags whitelist after synchronizing repository
doesn't affect already synchronized content
```
so probably you need
```
create_without_tag
sync
assert none
add tag
read
assert none
```

but you did:
```
create_without_tag
sync
assert none
add tag
sync
read
assert something
```
which is different thing

Or I just missed something

Reply:
The main idea of test is to verify that if I synchronize repo (= pull in some tags), add whitelist tag and re-synchronize, non-whitelisted tags are still present. This is by design and how feature is supposed to work.

Whitelist tags field takes effect only during sync. That's why second sync is needed.

Note that I have two assertions - one for field content and one for count of synchronized content. Assertion for field content changes, but that for content count does not. At the end of test, even though I only whitelisted one tag, I still expect at least two (because there was more than one already after first sync).

I could do:
```
create_without_tag
sync
assert whitelist field is empty
assert content count
add tag
read
assert whitelist field is not empty
sync
read
assert content count
```

But grouping assertions together feels more natural for me.

I updated docstring in 1446d11. Hopefully it does better job at explaining the intent.